### DISCUSSION TITLE
Reading IMU data with BLE - Getting duplicate entries...

### DISCUSSION THREAD

**Discussion by Lazerkill** [hbldh/bleak#1261] [Q&A] [answered]
Howdy,

I'm using Bleak and an Arduino Nano 33 BLE Sense (Rev2) to record gyro/accelerometer data and then send that data over BLE to my PC. I'm noticing an abnormality in my data... I'm consistently seeing duplicate entries across timestamps and gyro/accelerometer readings. See below: 
![image](https://user-images.githubusercontent.com/52724286/226212320-846b6683-b635-4060-b3cd-46957e816f13.png)

I'm fairly confident with my Arduino code since I don't see this same phenomenon when I print the Arduino output to serial. However, this is my first time working with Bleak so I'm wondering if that's where I've gone awry. Here's my Python code with Bleak:

```python
from bleak import BleakClient
import asyncio
from datetime import datetime


ADDRESS    = "61:81:C8:F7:EB:71"
GYRO_UUID  = "ed01c340-ff5f-41e2-8779-70fa0a22f891"
ACCEL_UUID = "ed01c340-ff5f-41e2-8779-70fa0a22f892"

# Create a handler for gyroscope notifications.
global gyro_byte_buffer; gyro_byte_buffer = []
def gyro_handler(characteristic, data):
    current_timestamp_entry = datetime.now()
    list_entry = (str(current_timestamp_entry) + "; " + data.hex())
    gyro_byte_buffer.append(list_entry)
       
# Create a handler for accelerometer notifications.
global accel_byte_buffer; accel_byte_buffer = []
def accel_handler(characteristic, data):
    current_timestamp_entry = datetime.now()
    list_entry = (str(current_timestamp_entry) + "; " + data.hex())
    accel_byte_buffer.append(list_entry)
    
# Connect to the Arduino and start reading data.
async def collect_data(address, gyro_uuid, accel_uuid):
    async with BleakClient(address) as client:
        print(f"Connected: {client.is_connected}")
        await client.start_notify(gyro_uuid, gyro_handler)
        await client.start_notify(accel_uuid, accel_handler)
        print("Notifications enabled... Reading data")
        await asyncio.sleep(15.0)
        await client.stop_notify(gyro_uuid)
        await client.stop_notify(accel_uuid)
        print("Reading complete")

if __name__ == "__main__":
    # Collect data over BLE.
    asyncio.run(collect_data(ADDRESS, GYRO_UUID, ACCEL_UUID))
    
    import struct
    import pandas as pd
    import time
    
    # Convert gyro data to a dataframe.
    timestamp_list = []
    x_list, y_list, z_list = [], [], []
    for entry in gyro_byte_buffer:
        [timestamp_entry, gyro_entry] = entry.split("; ")
        timestamp_entry = datetime.strptime(timestamp_entry, "%Y-%m-%d %H:%M:%S.%f")
        timestamp_entry = timestamp_entry.strftime("%H:%M:%S.%f")
        timestamp_list.append(timestamp_entry)
        
        gyro_entry = bytes.fromhex(gyro_entry)
        [x, y, z] = struct.unpack('fff', gyro_entry)
        x_list.append(x)
        y_list.append(y)
        z_list.append(z)  
    df_gyro = pd.DataFrame()
    df_gyro["Timestamp"] = timestamp_list
    df_gyro["X-axis (deg/s)"] = x_list
    df_gyro["Y-axis (deg/s)"] = y_list
    df_gyro["Z-axis (deg/s)"] = z_list
    
    # Convert accelerometer data to a dataframe.
    timestamp_list = []
    x_list, y_list, z_list = [], [], []
    for entry in accel_byte_buffer:
        [timestamp_entry, accel_data] = entry.split("; ")
        timestamp_entry = datetime.strptime(timestamp_entry, "%Y-%m-%d %H:%M:%S.%f")
        timestamp_entry = timestamp_entry.strftime("%H:%M:%S.%f")
        timestamp_list.append(timestamp_entry)
        
        accel_data = bytes.fromhex(accel_data)
        [x, y, z] = struct.unpack('fff', accel_data)
        x_list.append(x)
        y_list.append(y)
        z_list.append(z)  
    df_accel = pd.DataFrame()
    df_accel["Timestamp"] = timestamp_list
    df_accel["X-axis (g)"] = x_list
    df_accel["Y-axis (g)"] = y_list
    df_accel["Z-axis (g)"] = z_list
    
    # Merge gyro datafram and accelerometer dataframe into a single dataframe.
    df_merge = pd.merge(df_gyro, df_accel,  how="inner", on=["Timestamp", "Timestamp"], sort=True)  
    df_merge.index.name = 'Sample'
    
    # Create a column for "time delta".
    df_merge["Timestamp"] = pd.to_datetime(df_merge["Timestamp"])
    init_time = df_merge["Timestamp"].iloc[0]
    time_deltas = df_merge["Timestamp"] - init_time + pd.Timestamp(0)
    df_merge["Timedelta"] = time_deltas
    
    # Reorganzie columns.
    df_merge = df_merge[[
        "Timestamp",
        "Timedelta",
        "X-axis (deg/s)",
        "Y-axis (deg/s)",
        "Z-axis (deg/s)",
        "X-axis (g)",
        "Y-axis (g)",
        "Z-axis (g)"]].copy()
    
    # Write to Excel
    writer = pd.ExcelWriter("output.xlsx", datetime_format='hh:mm:ss.000')
    df_merge.to_excel(writer, sheet_name="Data")
    writer.sheets['Data'].set_column(0, 0, 8)
    writer.sheets['Data'].set_column(1, len(df_merge.columns)+1, 12)
    writer.close()
    print("Output to Excel complete")
```

Is there anything wrong with my implementation of Bleak that would cause this duplication to occur? I've attached the Excel output below.

Thanks in advance and thanks for making this wonderful library! 

[Bleak Output.xlsx](https://github.com/hbldh/bleak/files/11012428/Bleak.Output.xlsx)
[Bleak Output.mp4](https://user-images.githubusercontent.com/52724286/226212745-5acd08db-41fb-44f5-b690-9c56786ba0ed.mp4)
(The duplication of data is really apparent in the animation of the data above)

**Comment by dlech** [comment, marked as answer]
Operating system?

Bluetooth packet logs?

**Reply by Lazerkill** [reply]
Hi! Sorry I should've included this from the start:
Bleak version: 0.20.0
Python version: 3.10.9
Operating System: Windows10 Home
BLE: Arduino Nano 33 BLE Sense (Rev2) 

What's the best way to check for packet loss in Bleak? Apologies for the dumb question. Still very new to Bleak.

**Reply by dlech** [reply]
https://bleak.readthedocs.io/en/latest/troubleshooting.html#windows-10

**Reply by Lazerkill** [reply]
Thank you so much for the link! Insofar as I can tell, everything seems correct. I'm getting a very consistent stream of 24 byte packets coming in from my Arduino. I'm using the gyro characteristic to send 3 floats (4 bytes per float) and the accelerometer characteristic to send 3 floats (4 bytes per float). 

Am I missing something here?

![image](https://user-images.githubusercontent.com/52724286/226216075-cf866952-3a9c-4703-81d0-aceb5512a4b5.png)

![image](https://user-images.githubusercontent.com/52724286/226216685-bb83b4e9-3b64-4c4e-b296-18a1c315b96e.png)

[wireshark_output.pcapng.gz](https://github.com/hbldh/bleak/files/11012550/wireshark_output.pcapng.gz)

**Reply by dlech** [reply]
Are you sure the duplication doesn't come from `pb.merge()`?

**Reply by Lazerkill** [reply]
Good catch! I added new sheets to my Excel export that include the gyro data and accel data prior to merge. You can see the merged set has instances of duplication but the unmerged data does not...
[output.xlsx](https://github.com/hbldh/bleak/files/11130846/output.xlsx)
![duplicate](https://user-images.githubusercontent.com/52724286/229323363-32bf4cad-3310-4aed-87d2-8153977fd8de.PNG)
![no duplicate](https://user-images.githubusercontent.com/52724286/229323369-f1b12674-6ec8-475a-a2ae-10ed73115ad7.PNG)
By setting up different tabs I was also able to realize that merging the data with Pandas is robbing me of a significant number of packets of data (making the data rate look worse than it really is). Looks like I need to think of a better way to merge these sets of data...

Nonetheless, **thank you for the help**. Going to mark this as the answer 👍 

Unrelated question, (and maybe I should take this to the Arduino forums instead) but would you know why the data rate would be so low (even before the merge takes place)? Looking at the individual gyro and accelerometer dataframes, I'm getting a total of 433 packets that are 12 bytes long each over the course of 10 seconds... By my math that equates to ~520 bytes per second. That's a fraction of the 2 megabits per second that this BLE 5.0 is advertised to carry. I know the IMU isn't the bottleneck because I was getting 100Hz when printing to serial terminal over USB.

**Reply by dlech** [reply]
There are lots of factors that affect throughput. It really helps to have a sniffer to understand what is actually going over the air and when. PunchThrough has a nice [blog series](https://punchthrough.com/maximizing-ble-throughput-part-3-data-length-extension-dle-2) on the topic.
